src/ff/prime_field.rs

```diff
@@ -0,0 +1,307 @@
+        /// is true provided that `PRIME` is kept to at most 64 bits in value.
+        ///
+        /// This method is simpler than rejection sampling for these small prime fields.
+        impl<T: Into<u128>> From<T> for $field {
+            fn from(v: T) -> Self {
+                #[allow(clippy::cast_possible_truncation)]
+                Self((v.into() % u128::from(Self::PRIME)) as <Self as Field>::Integer)
+            }
+        }
+
+        impl Debug for $field {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                write!(f, "{}_mod{}", self.0, Self::PRIME)
+            }
+        }
+    )* };
+}
+
+#[derive(Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub struct Fp2(<Self as Field>::Integer);
+
+impl Field for Fp2 {
+    type Integer = u8;
+    const PRIME: Self::Integer = 2;
+    const ZERO: Self = Fp2(0);
+    const ONE: Self = Fp2(1);
+}
+
+impl BinaryField for Fp2 {}
+
+impl BitAnd for Fp2 {
+    type Output = Self;
+
+    fn bitand(self, rhs: Self) -> Self::Output {
+        Self((self.0 & rhs.0) & 1)
+    }
+}
+
+impl BitAndAssign for Fp2 {
+    fn bitand_assign(&mut self, rhs: Self) {
+        *self = *self & rhs;
+    }
+}
+
+impl BitOr for Fp2 {
+    type Output = Self;
+
+    fn bitor(self, rhs: Self) -> Self::Output {
+        Self((self.0 | rhs.0) & 1)
+    }
+}
+
+impl BitOrAssign for Fp2 {
+    fn bitor_assign(&mut self, rhs: Self) {
+        *self = *self | rhs;
+    }
+}
+
+impl BitXor for Fp2 {
+    type Output = Self;
+
+    fn bitxor(self, rhs: Self) -> Self::Output {
+        Self((self.0 ^ rhs.0) & 1)
+    }
+}
+
+impl BitXorAssign for Fp2 {
+    fn bitxor_assign(&mut self, rhs: Self) {
+        *self = *self ^ rhs;
+    }
+}
+
+impl Not for Fp2 {
+    type Output = Self;
+
+    fn not(self) -> Self::Output {
+        Self((self.0 + 1) & 1)
+    }
+}
+
+#[derive(Clone, Copy, PartialEq)]
+#[cfg_attr(feature = "enable-serde", derive(Serialize, Deserialize))]
+pub struct Fp31(<Self as Field>::Integer);
+
+impl Field for Fp31 {
+    type Integer = u8;
+    const PRIME: Self::Integer = 31;
+    const ZERO: Self = Fp31(0);
+    const ONE: Self = Fp31(1);
+}
+
+field_impl! { Fp2, Fp31 }
+
+#[cfg(test)]
+mod test {
+    use super::{Field, Fp2, Fp31};
+    use std::ops::Mul;
+
+    #[test]
+    fn fp2() {
+        let x = Fp2::from(false);
+        let y = Fp2::from(true);
+
+        assert_eq!(Fp2(1), x + y);
+        assert_eq!(Fp2(0), x * y);
+        assert_eq!(Fp2(1), x - y);
+
+        let mut x = Fp2(1);
+        x += Fp2(1);
+        assert_eq!(Fp2(0), x);
+    }
+
+    #[test]
+    fn fp2_binary_op() {
+        let zero = Fp2::ZERO;
+        let one = Fp2::ONE;
+
+        assert_eq!(one, one & one);
+        assert_eq!(zero, zero & one);
+        assert_eq!(zero, one & zero);
+        assert_eq!(zero, zero & zero);
+        assert_eq!(zero, Fp2(31) & Fp2(32));
+        assert_eq!(one, Fp2(31) & Fp2(63));
```

Comment:
That's interesting. Is this the behavior we want? Do we want it to assert that the value is either 0 or 1 in the constructor? I don't see a ::new() method at all. Does that mean that if you do not do any operation (like this `&` operation here) that it would remain as 31?

Reply:
Yes, that is correct. Since Fp2 implements the `Field` trait, internal value is `u8` and it supports `From<u128>`. Implementing `::new()` doesn't help much because one could still use `Fp2::from()`. That said, this comes in handy when a boolean circuit protocol executes e.g. Sum on $Z_2$. We can do something like `Fp2::from(b_1 + ... + b_n)`.

Reply:
Sure, but shouldn’t the Fp2::from() implement some kind of range checking and/or casting?

Reply:
`From<u128>` implementation is inside the macro and common across `Field` (lines 86-91). We could remove it from the macro and force each `Fp` structs to do custom implementations.

I think it makes sense to not do range checking, though. We want to accept any range of numbers to `FpX::from()` so that we don't have to worry about the input going out of the range, but internally modulo takes care of overflows. It may turn out that we always want to force p or 1 for `Fp2`, and if so, we can do custom `From<u128>` implementations later?

Reply:
As @martinthomson points out, "Fp31 clamps in it’s From impl". Indeed, on lines 86-91 there is a "` mod self::PRIME`. That should work just fine as a clamping mechanism. I think that's a perfectly reasonable way to ensure that the internally stored byte is always either a 0 or a 1. Let's do that, so that we do not need a `& 1` after every operation.

Reply:
To that end, I would not use `Fp2(x)` in tests as that creates an instance of the object that violates the invariant that Fp2 carries only 0 or 1.  I know that this is intended to break that deliberately, but that's not really a fair test.

Reply:
I agree with @martinthomson

Reply:
Oh, I missed that clamp part. Let me remove `& 1` and use `::from()` everywhere.